### PR TITLE
remote/s3: s3 remote state storage support

### PR DESCRIPTION
**Pull request by kubek2k**
fixes: https://github.com/hashicorp/terraform/issues/880

### CODE REVIEW

File: command/remote.go [hunk 1]
```diff
@@ -57,10 +60,13 @@ func (c *RemoteCommand) Run(args []string) int {
...
-		"address":      address,
-		"access_token": accessToken,
-		"name":         name,
-		"path":         path,
+		"address":        address,
+		"access_token":   accessToken,
+		"security_token": securityToken,
+		"name":           name,
+		"path":           path,
+		"bucket":         bucket,
```
Comment (armon) [inline command/remote.go]: I would just make bucket part of path. e.g. the path should be specified as s3://bucket/prefix/

File: remote/s3.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	err := c.Bucket.Put(c.Path, state, "application/json", s3.Private, options)
+	switch err.(type) {
+	case *s3.Error:
+		s3Err := err.(*s3.Error)
+
+		// Handle the error codes
+		switch s3Err.StatusCode {
+		case http.StatusOK:
+			return nil
+		case http.StatusConflict:
```
Comment (armon) [inline remote/s3.go]: S3 should not return ErrConflict or ErrServerNewer. Those semantics have to do with enforcement of the state serial number.

File: command/remote.go (hunk 1 above)
Comment (kubek2k) [inline command/remote.go]: well - I wasn't sure what should be where - :+1:

File: remote/s3.go (hunk 2 above)
Comment (kubek2k) [inline remote/s3.go]: thats a dumb code copied from Atlas implementation.
Should it be removed from there as well @armon ?

File: command/remote.go (hunk 1 above)
Comment (kubek2k) [inline command/remote.go]: @armon - maybe it would be better to use url param for this instead ?

Comment (armon) [inline command/remote.go]: Agreed. We can probably overload `address` for it.

File: remote/s3.go (hunk 2 above)
Comment (armon) [inline remote/s3.go]: @kubek2k Atlas actually does implement those features, it's part of the value add :)

Comment (kubek2k) [inline remote/s3.go]: ah - but I couldn't find any version being set in the headers sent along the PUT request - is it some kind of other mechnism?

Comment (armon) [inline remote/s3.go]: Yes the server parses the actual contents of the file.

Comment (kubek2k) [inline remote/s3.go]: ah got it :+1: 

so if I would like to make a conditional PUT instead for S3, I would have to store version someplace aside (as in interface we only get state as byte array)? Assuming other stores maybe version should come as an addtional arg?

**Comment by kubek2k** [comment]
@armon take a look now

**Comment by armon** [comment]
@kubek2k We'll try to take a look at this for Terraform 0.4. Thanks!
/cc @phinze

**Comment by radeksimko** [comment]
Few notes/suggestions:

1. I feel that the whole remote storage feature will need proper support for file configuration as it's becoming more and more confusing while mixing all the backend options together in the same one-level namespace (Consul, Atlas, HTTP)
2. Regarding S3, you'll probably face the same issue with credentials which I'm trying to solve in https://github.com/hashicorp/terraform/pull/1049

**Comment by mitchellh** [comment]
Sorry @kubek2k but could you re-implement this according to the new remote interface? it should be _much_ simpler now and we give you acceptance tests out of the box with `testClient`! See `state/remote`

**Comment by apparentlymart** [comment]
This has been updated to the new API in PR #1723

**Pull request closed without merging**